pyvrp/IteratedLocalSearch.py

```diff
@@ -0,0 +1,162 @@
+
+
+@dataclass
+class IteratedLocalSearchParams:
+    """
+    Parameters for the iterated local search algorithm.
+    """
+
+    pass
+
+
+class IteratedLocalSearch:
+    """
+    Creates an IteratedLocalSearch instance.
+
+    Parameters
+    ----------
+    data
+        The problem data instance.
+    penalty_manager
+        Penalty manager to use.
+    rng
+        Random number generator.
+    perturb_method
+        Perturb method to use.
+    search_method
+        Search method to use.
+    acceptance_criterion
+        Acceptance criterion to use.
+    params
+        Iterated local search parameters to use. If not provided, a default
+        will be used.
+    """
+
+    def __init__(
+        self,
+        data: ProblemData,
+        penalty_manager: PenaltyManager,
+        rng: RandomNumberGenerator,
+        perturb_method: SearchMethod,
+        search_method: SearchMethod,
+        acceptance_criterion: AcceptanceCriterion,
+        params: IteratedLocalSearchParams,
+    ):
+        self._data = data
+        self._pm = penalty_manager
+        self._rng = rng
+        self._perturb = perturb_method
+        self._search = search_method
+        self._accept = acceptance_criterion
+        self._params = params
+
+    @property
+    def _cost_evaluator(self) -> CostEvaluator:
+        return self._pm.cost_evaluator()
+
+    def _stats(self, solution: Solution) -> tuple[float, bool]:
+        cost = self._cost_evaluator.penalised_cost(solution)
+        is_feas = solution.is_feasible()
+        return cost, is_feas
+
+    def run(
+        self,
+        stop: StoppingCriterion,
+        initial_solution: Solution,
+        collect_stats: bool = True,
+        display: bool = False,
+    ) -> Result:
+        """
+        Runs the iterated local search algorithm.
+
+        Parameters
+        ----------
+        stop
+            Stopping criterion to use. The algorithm runs until the first time
+            the stopping criterion returns ``True``.
+        initial_solution
+            The initial solution to use in the first iteration.
+        collect_stats
+            Whether to collect statistics about the solver's progress. Default
+            ``True``.
+        """
+        print_progress = ProgressPrinter(should_print=display)
+        print_progress.start(self._data)
+        conv_manager = ConvergenceManager(initial_num_destroy=30)
+
+        start = time.perf_counter()
+        stats = Statistics(collect_stats=collect_stats)
+        iters = 0
+        best = current = initial_solution
+
+        while not stop(self._cost_evaluator.cost(best)):
+            iters += 1
+
+            perturbed = self._perturb(
+                current, self._cost_evaluator, conv_manager.num_destroy
+            )
+            diff = different_neighbours(current, perturbed)
+            cost_evaluator = CostEvaluator([20], 6, 6)
+            candidate = self._search(perturbed, cost_evaluator, diff)
```

Comment:
So this doesn't work yet because the cost evaluator penalties do not work for all instances.

Reply:
When to increase penalties?